File: netlogo-gui/src/main/nvm/Reporter.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.nvm;

import org.nlogo.agent.Agent;
import org.nlogo.agent.AgentSet;
import org.nlogo.api.LogoException;

public abstract strictfp class Reporter
    extends Instruction {
  public abstract Object report(final Context context)
      throws LogoException;

  public final void checkAgentClass(Agent agent, Context context)
      throws LogoException {
    if ((agent.getAgentBit() & agentBits) == 0) {
      throwAgentClassException(context, agent.kind());
    }
  }

  public final void checkAgentSetClass(AgentSet agents, Context context)
      throws LogoException {
    if ((agents.getAgentBit() & agentBits) == 0) {
      throwAgentClassException(context, agents.kind());
    }
  }

}

File: netlogo-gui/src/main/nvm/Reporter.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.nvm

import org.nlogo.api.LogoException
import org.nlogo.agent.{ Agent, AgentSet }

abstract class Reporter extends Instruction {
  @throws(classOf[LogoException])
  def report(context: Context): AnyRef

  @throws(classOf[LogoException])
  final def checkAgentClass(agent: Agent, context: Context): Unit = {
    if ((agent.getAgentBit & agentBits) == 0) {
      throwAgentClassException(context, agent.kind)
    }
  }

  @throws(classOf[LogoException])
  final def checkAgentSetClass(agents: AgentSet, context: Context): Unit = {
    if ((agents.getAgentBit & agentBits) == 0) {
      throwAgentClassException(context, agents.kind)
    }
  }
}
